<p>Tag-to-sample file (CSV)</p>
<input type="text" name="tags" class="input_file csv tags" value="">
<div class="demux_warnings"></div>

<p>Inputs R1/R2 by library</p>
<div class="illumina_reads">

</div>

<br>
<p>Primers file (IUPAC supported)</p>
<input type="text" name="primers" class="input_file fasta primers" value="">

<br>
<p class="no_vertical_margins"><input type="checkbox" class="param_value trim_end" name="trim_end">Trim also the primer sequence at the end of the read (if present)</p>
<br>
<p> Filtering by length the trimmed read (discarded ones will be in empty.fasta)<a href="" download=""><img src="/imgs/download.png" class="download"></a></p>
<input type="text" name="trim_length" class="param_value number integer" value="1">

<h4>Outputs</h4>

<p>Two output files will be created for each sample in the CSV file.
They will be named run_sample_fwd.fastq and run_sample_rev.fastq.</p>

<div class="file_list" readonly></div>

<div class="options">

	<br>
	<p><input type="checkbox" name="mistags" class="param_value">Output undertermined reads (no tags retrieved)<a href="" download=""><img src="/imgs/download.png" class="download"></a></p>
	<br>
	<p>Mismatch allowed in primers: <input type="text" name="errors" class="param_value number integer" value=0></p>
</div>
